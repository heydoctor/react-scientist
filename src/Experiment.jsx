// @flow

import React from 'react';
import crc32 from 'fbjs/lib/crc32';
import Cookie from 'js-cookie';

type Variant = {
  name: string,
  weight: number,
  render: Function,
};

type VariantList = Array<Variant>;

type ExperimentProps = {
  name: string,
  id?: string | number,
  userId?: string,
  variants: VariantList,
};

type StartArgs = {
  experimentName: string,
  experimentId?: string | number,
  variantIndex: number,
  variantName: string,
};

export const varianceHelpers = {
  /**
   * Gets the sum of all weights after conversion from percent to integer
   */
  getWeightSum(variants: VariantList): number {
    return variants.reduce((sum, v) => {
      if (v.weight > 1) {
        throwError(
          'Weights should be < 1 and in decimal form. The sum of the variant weights should most commonly add up to 1, e.g. 0.5 & 0.5'
        );
      }
      return sum + v.weight * 100;
    }, 0);
  },

  /**
   * Create an array filled with variant indexes proportionate to their weight
   * Given variants: [{ name: 'One', weight: 0.2 }, { name: 'Two', weight: 0.8 }]
   * Distribtuion array would consist of 20 "One" elements and 80 "Two" elements
   */
  getWeightedDistribution(variants: VariantList): Array<number> {
    return variants.reduce(
      (arr, { weight }, idx) => [...arr, ...Array(weight * 100).fill(idx)],
      []
    );
  },

  /**
   * Random number between 0 and 100. If userId is provided, calculate fixed number within
   * bounds to maintain consistent experience across repeat visits
   */
  getWeightedIndex(weightSum: number, staticId: number): number {
    return staticId
      ? Math.abs(crc32(`${staticId}`) % weightSum)
      : Math.floor(Math.random() * weightSum);
  },

  /**
   * Chooses a variant
   */
  selectVariantIndex(
    variants: VariantList,
    staticId: ?number | ?string
  ): number {
    const weightSum = this.getWeightSum(variants);
    const weightedDistribution = this.getWeightedDistribution(variants);
    const weightedIndex = this.getWeightedIndex(weightSum, staticId);
    const variantIndex = weightedDistribution[weightedIndex];
    return variantIndex;
  },
};

const throwError = message => {
  throw new Error(`<Experiment />: ${message}`);
};

class Experiment extends React.Component<ExperimentProps> {
  static onStart = (args: StartArgs) => {
    throwError(
      'This method should be overwritten somewhere before React is mounted, e.g. Experiment.onStart = () => {...}'
    );
  };

  storeKey = `scientist.${
    this.props.userId !== undefined ? this.props.userId : 'guest'
  }.${this.props.name}`;

  getVariant = () => {
    const { name, variants, id, userId } = this.props;

    const storedVariantIndex = Cookie.get(this.storeKey);
    if (storedVariantIndex !== undefined && storedVariantIndex !== null)
      return variants[storedVariantIndex];

    const variantIndex = varianceHelpers.selectVariantIndex(variants, userId);
    const selectedVariant = variants[variantIndex];
    Cookie.set(this.storeKey, variantIndex);

    this.constructor.onStart({
      experimentName: name,
      experimentId: id,
      variantIndex,
      variantName: selectedVariant.name,
    });

    return selectedVariant;
  };

  render() {
    const variant = this.getVariant();
    if (variant) return variant.render();

    return null;
  }
}

export default Experiment;
