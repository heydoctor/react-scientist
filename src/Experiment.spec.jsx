import React from 'react';
import { render, cleanup } from 'react-testing-library';
import Experiment, { varianceHelpers } from './Experiment';
import Storage from './storage';

const variants = [
	{
		name: 'Control',
		weight: 0.5,
		render: () => <div>Control</div>,
	},
	{
		name: 'Test',
		weight: 0.5,
		render: () => <div>Test</div>,
	},
];

const Subject = props => (
	<Experiment variants={props.variants || variants} {...props} />
);

afterEach(cleanup);

describe('varianceHelpers', () => {
	describe('#getWeightSum', () => {
		test('throws if any weight is greater than 1', () => {
			expect(() =>
				varianceHelpers.getWeightSum([{ name: 'Wrong', weight: 2 }])
			).toThrowErrorMatchingSnapshot();
		});

		test('returns a sum of all normalized weights', () => {
			const sum = varianceHelpers.getWeightSum(variants);
			expect(sum).toEqual(100);
		});
	});

	describe('#getWeightedDistribution', () => {
		test('returns an array with variant indexes as values, repeated proportionate to their weight', () => {
			const dist = varianceHelpers.getWeightedDistribution([
				{ weight: 0.3 },
				{ weight: 0.7 },
			]);
			const counts = {};
			dist.forEach(d => {
				const index = `${d}`;
				counts[d] = typeof counts[index] === 'number' ? counts[index] + 1 : 1;
			});
			expect(counts[0]).toEqual(30);
			expect(counts[1]).toEqual(70);
		});
	});

	describe('#getWeightedIndex', () => {
		test('returns a number in between 0 and weightSum', () => {
			const weightSum = 100;
			const index = varianceHelpers.getWeightedIndex(weightSum);
			expect(index).toBeGreaterThanOrEqual(0);
			expect(index).toBeLessThan(weightSum);
		});

		test('always returns same index when passed static id', () => {
			let index;
			const staticId = 42;
			for (let i; i < 500; i++) {
				const nextIndex = varianceHelpers.getWeightedIndex(variants, staticId);
				if (index !== nextIndex) {
					throw new Error(
						'`getWeightedIndex` returned a different value when passed a static id'
					);
				}
				if (!index) {
					index = nextIndex;
				}
			}
		});
	});

	describe('#selectVariantIndex', () => {
		test('returns an valid variant index', () => {
			const index = varianceHelpers.selectVariantIndex(variants);
			expect(typeof index).toEqual('number');
			expect([0, 1]).toContain(index);
		});
	});
});

test('Experiment#onStart throws', () => {
	expect(() => Experiment.onStart()).toThrowErrorMatchingSnapshot();
});

describe('<Experiment />', () => {
	beforeAll(() => {
		Experiment.onStart = jest.fn();
	});

	afterEach(() => {
		Experiment.onStart.mockReset();
	});

	test('storeKey is set to guest if no userId is provided', () => {
		const name = 'Test';
		const { container } = render(<Subject name={name} />);
		expect(Storage.getItem(`scientist.guest.${name}`)).toBeDefined();
	});

	test('storeKey uses userId if provided', () => {
		const name = 'Test';
		const userId = 3;
		const { container } = render(<Subject name={name} userId={userId} />);
		expect(Storage.getItem(`scientist.${userId}.${name}`)).toBeDefined();
	});

	test('calls Experiment.onStart only once', () => {
		const name = 'onStart';
		const { rerender } = render(<Subject name={name} />);
		expect(Experiment.onStart).toHaveBeenCalledTimes(1);
		rerender(<Subject name={name} />);
		expect(Experiment.onStart).toHaveBeenCalledTimes(1);
	});

	test('renders one of the variants', () => {
		const name = 'Render';
		const { container } = render(<Subject name={name} />);

		expect(['Control', 'Test']).toContain(container.textContent);
	});
});
