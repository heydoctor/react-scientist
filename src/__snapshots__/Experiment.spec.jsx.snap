// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Experiment#onStart throws 1`] = `"<Experiment />: This method should be overwritten somewhere before React is mounted, e.g. Experiment.onStart = () => {...}"`;

exports[`varianceHelpers #getWeightSum throws if any weight is greater than 1 1`] = `"<Experiment />: Weights should be < 1 and in decimal form. The sum of the variant weights should most commonly add up to 1, e.g. 0.5 & 0.5"`;
